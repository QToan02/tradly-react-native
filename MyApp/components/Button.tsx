import { Pressable, PressableProps, StyleSheet, Text } from "react-native";

export interface MyButtonProps extends PressableProps {
  title: string;
  onPress: () => void;
}

const MyButton = ({ title, onPress, ...rest }: MyButtonProps) => {
  return (
    <Pressable style={styles.btn} onPress={onPress} {...rest}>
      <Text>{title}</Text>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  btn: {
    borderColor: "blue",
    borderWidth: 1,
    padding: 10,
    borderRadius: 5,
  },
});

export default MyButton;
