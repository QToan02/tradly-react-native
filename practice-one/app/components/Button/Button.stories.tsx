import { Image } from 'react-native'
import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { Button } from '../index'

export default {
  title: 'Button',
  component: Button,
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

const LeftIcon = (
  <Image
    // eslint-disable-next-line global-require
    source={require('@assets/icon.png')}
    fadeDuration={0}
    style={{ width: 24, height: 24 }}
  />
)

const RightIcon = (
  <Image
    // eslint-disable-next-line global-require
    source={require('@assets/favicon.png')}
    fadeDuration={0}
    style={{ width: 24, height: 24 }}
  />
)

export const Default = Template.bind({})
Default.args = {
  title: 'My Button',
}

export const SecondaryButton = Template.bind({})
SecondaryButton.args = {
  title: 'My Button',
  variant: 'secondary',
}

export const OutlineButton = Template.bind({})
OutlineButton.args = {
  title: 'My Button',
  variant: 'tertiary',
}

export const FullButtonWithoutRadius = Template.bind({})
FullButtonWithoutRadius.args = {
  title: 'My Button',
  variant: 'quaternary',
}

export const SmallButton = Template.bind({})
SmallButton.args = {
  title: 'My Button',
  shrink: true,
}

export const ButtonWithLeftIcon = Template.bind({})
ButtonWithLeftIcon.args = {
  leftIcon: LeftIcon,
  variant: 'secondary',
  title: 'My Icon Button',
}

export const ButtonWithRightIcon = Template.bind({})
ButtonWithRightIcon.args = {
  rightIcon: RightIcon,
  title: 'My Icon Button',
}
