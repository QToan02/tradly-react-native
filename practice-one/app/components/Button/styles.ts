import { StyleSheet } from 'react-native'

import { COLORS, FONT_FAMILY } from '@constants'

const styles = StyleSheet.create({
  btn: {
    paddingVertical: 13,
    borderRadius: 50,
    borderWidth: 2,
    borderColor: COLORS.PRIMARY,
    textTransform: 'capitalize',
  },
  btnSmall: {
    alignSelf: 'flex-start',
    paddingVertical: 7,
    paddingHorizontal: 20,
  },
  primary: {
    backgroundColor: COLORS.PRIMARY,
    color: COLORS.WHITE,
  },
  secondary: {
    backgroundColor: COLORS.SECONDARY,
    color: COLORS.PRIMARY,
    borderColor: COLORS.TRANSPARENT,
  },
  tertiary: {
    backgroundColor: COLORS.TRANSPARENT,
    color: COLORS.WHITE,
    borderColor: COLORS.WHITE,
  },
  quaternary: {
    backgroundColor: COLORS.WHITE,
    color: COLORS.BLACK,
    borderRadius: 0,
  },
  btnContent: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    columnGap: 6,
  },
  btnTitle: {
    fontFamily: FONT_FAMILY.MONTSERRAT[500],
    textAlign: 'center',
  },
})

export default styles
