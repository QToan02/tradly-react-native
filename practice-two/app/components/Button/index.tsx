import { memo } from 'react'
import { Image, ImageSourcePropType } from 'react-native'
import isEqual from 'react-fast-compare'

import { imageStyles } from '@styles'

import ButtonStyled, { ButtonProps as StyledButtonProps } from './styles'

export type ButtonProps = {
  title: string
  leftIcon?: ImageSourcePropType
  rightIcon?: ImageSourcePropType
} & StyledButtonProps

const Button = ({ title, leftIcon, rightIcon, ...rest }: ButtonProps) => {
  return (
    <ButtonStyled {...rest}>
      {leftIcon && (
        <ButtonStyled.Icon>
          <Image source={leftIcon} style={imageStyles.icon} />
        </ButtonStyled.Icon>
      )}
      <ButtonStyled.Text>{title}</ButtonStyled.Text>
      {rightIcon && (
        <ButtonStyled.Icon>
          <Image source={rightIcon} style={imageStyles.icon} />
        </ButtonStyled.Icon>
      )}
    </ButtonStyled>
  )
}

export default memo(Button, isEqual)
