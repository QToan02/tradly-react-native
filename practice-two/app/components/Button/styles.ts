import {
  ButtonIcon,
  GetProps,
  SizableText,
  Stack,
  createStyledContext,
  getTokens,
  styled,
  withStaticProperties,
} from 'tamagui'

const ButtonContext = createStyledContext({
  color: undefined,
  ellipse: undefined,
  fontFamily: undefined,
  fontSize: undefined,
  fontStyle: undefined,
  fontWeight: undefined,
  letterSpacing: undefined,
  maxFontSizeMultiplier: undefined,
  size: undefined,
  textAlign: undefined,
})

const CustomButtonFrame = styled(Stack, {
  name: 'Button',
  context: ButtonContext,
  paddingVertical: '$space.3',
  borderRadius: '$radius.12',
  borderWidth: '$space.1',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'center',
  alignItems: 'center',
  columnGap: getTokens().space[2].val - 1,
  pressStyle: {
    opacity: 0.5,
  },

  variants: {
    variant: {
      primary: {
        borderColor: '$color.primary',
        backgroundColor: '$color.primary',
        color: '$color.white',
      },

      secondary: {
        borderColor: '$color.transparent',
        backgroundColor: '$color.secondary',
        color: '$color.primary',
      },

      tertiary: {
        borderColor: '$color.white',
        backgroundColor: '$color.transparent',
        color: '$color.white',
      },

      quaternary: {
        borderRadius: '$radius.0',
        borderColor: '$color.white',
        backgroundColor: '$color.white',
        color: '$color.black',
      },
    },

    shrink: {
      true: {
        alignSelf: 'flex-start',
        paddingVertical: '$space.2',
        paddingHorizontal: getTokens().space[4].val + 2,
      },
    },
  } as const,

  defaultVariants: {
    variant: 'primary',
  },
})

const ButtonText = styled(SizableText, {
  name: 'Button',
  context: ButtonContext,
  userSelect: 'none',
  size: '$2',
  fontWeight: '$2',
  textTransform: 'capitalize',
  textAlign: 'center',
})

const ButtonStyled = withStaticProperties(CustomButtonFrame, {
  Text: ButtonText,
  Icon: ButtonIcon,
  Props: ButtonContext.Provider,
})

export type ButtonProps = GetProps<typeof CustomButtonFrame>

export default ButtonStyled
