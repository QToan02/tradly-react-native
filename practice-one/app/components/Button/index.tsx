import { memo, useMemo } from 'react'
import isEqual from 'react-fast-compare'
import {
  Image,
  ImageProps,
  StyleProp,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle,
} from 'react-native'

import { TButtonVariant } from '@types'
import Paragraph from '@components/Paragraph'
import { imageStyles } from '@styles'
import { COLORS } from '@constants'

import styles from './styles'

export interface ButtonProps extends TouchableOpacityProps {
  title: string
  leftIcon?: ImageProps['source']
  rightIcon?: ImageProps['source']
  variant?: TButtonVariant
  shrink?: boolean
  style?: StyleProp<ViewStyle>
  titleStyle?: StyleProp<TextStyle>
  onPress: () => void
}

const Button = ({
  title,
  leftIcon,
  rightIcon,
  variant = 'primary',
  shrink = false,
  style,
  titleStyle,
  onPress,
  ...rest
}: ButtonProps) => {
  const titleVariantStyle = useMemo((): TextStyle => {
    switch (variant) {
      case 'primary':
        return { color: COLORS.WHITE }
      case 'secondary':
        return { color: COLORS.PRIMARY }
      case 'tertiary':
        return { color: COLORS.WHITE }
      case 'quaternary':
        return { color: COLORS.BLACK }
      default:
        return { color: COLORS.WHITE }
    }
  }, [variant])

  return (
    <TouchableOpacity
      style={[styles.btn, styles[variant], shrink && styles.btnSmall, style]}
      onPress={onPress}
      {...rest}
    >
      <View style={styles.btnContent}>
        {leftIcon && <Image source={leftIcon} style={imageStyles.icon} />}
        <Paragraph
          size="md"
          style={[titleVariantStyle, styles.btnTitle, titleStyle]}
          content={title}
        />
        {rightIcon && <Image source={rightIcon} style={imageStyles.icon} />}
      </View>
    </TouchableOpacity>
  )
}

export default memo(Button, isEqual)
