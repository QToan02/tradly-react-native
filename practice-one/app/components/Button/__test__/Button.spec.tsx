import { render, fireEvent } from '@testing-library/react-native'
import Button from '../index'

describe('Testing button component', () => {
  it('Render button component with primary variant', () => {
    const mockOnPress = jest.fn()
    const component = render(
      <Button title="press here" variant="primary" onPress={mockOnPress} />
    ).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Press on button with fire onPress event', () => {
    const mockOnPress = jest.fn()
    const { getByText } = render(<Button title="press here" onPress={mockOnPress} />)

    fireEvent.press(getByText('press here'))
    expect(mockOnPress).toBeCalledTimes(1)
  })
})
