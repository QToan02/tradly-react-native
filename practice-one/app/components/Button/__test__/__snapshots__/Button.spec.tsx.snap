// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing button component Render button component with primary variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#33907C",
      "borderColor": "#33907C",
      "borderRadius": 50,
      "borderWidth": 2,
      "color": "#fff",
      "opacity": 1,
      "paddingVertical": 13,
      "textTransform": "capitalize",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "columnGap": 6,
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_400Regular",
          },
          {
            "fontSize": 16,
          },
          [
            {
              "backgroundColor": "#33907C",
              "color": "#fff",
            },
            {
              "fontFamily": "Montserrat_500Medium",
              "textAlign": "center",
            },
          ],
        ]
      }
    >
      press here
    </Text>
  </View>
</View>
`;
